cnxdb/migrations/20170812000000_current-modules-view.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+
+
+def up(cursor):
+    """Replace current_modules with correct logic"""
+    cursor.execute("""
+DROP VIEW current_modules;
+CREATE OR REPLACE VIEW current_modules AS
+WITH latest_idents (module_ident) AS (
+          SELECT module_ident FROM modules m2 join modulestates ms
+                on m2.stateid = ms.stateid
+                WHERE m2.major_version = (
+                    SELECT max(major_version) FROM modules m3
+                        WHERE m2.uuid = m3.uuid
+                )
+                AND
+                (m2.minor_version IS NULL OR
+                 m2.minor_version = (
+                    SELECT max(minor_version) FROM modules m4
+                        WHERE m2.uuid = m4.uuid AND
+                        m2.major_version = m4.major_version
+                    )
+                )
```

Comment:
How about we add this here?

```sql
                AND
                (m2.module_ident = (
                    SELECT max(module_ident) FROM modules m5
                        WHERE m2.uuid = m5.uuid AND
                              m2.major_version = m5.major_version AND
                              m2.minor_version = m2.minor_version
                    )
                )
```

If there are multiple rows for the same version like on QA... :(, we'll use the one with the `max(module_ident)`?

Reply:
There should not be duplicated uuids like that. Nor duplicate uuid+version. I'd rather not enshrine it in the method, since if the data get's broken again, that will hide the breakage. I'm hoping that's just QA test-data wierdness. The subcol w/ same uuid as collection is probably from the broken subcol_uuid I just fixed

Reply:
We should probably add another migration that fixes the QA database, or run the hotfix script...?

Reply:
I think it's reasonable to hand-tweak the DB on QA, if we verify that we don't see that data-pattern on staging/production. I'm looking at it right now.

Reply:
Checking directly on QA for dups, with this code:

Reply:
Which takes a _Very_ long time, due to _on delete cascade_ constraints into the full-text-search index tables …

Reply:
o.k. added indexes to the foreign side of all the on-delete-cascade refs. Deletions now much faster. Deleted all the orphaned CompositeModules and SubCollections (both existed). Now there are no dups in datamigrations.corrected_latest

Reply:
(sigh- probably need to make a migration for adding all the needed indexes)

Reply:
Shouldn't `delete from modules m where portal_type = 'CompositeModule' and not exists (select 1 from trees t where t.documentid = m.module_ident);` be a migration?

Reply:
Can we modify https://github.com/Connexions/devops/blob/master/scripts/remove_duplicate_minor_version_books.py to delete duplicate composite modules?

Reply:
https://github.com/Connexions/devops/pull/16

Reply:
Migration completed successfully after running the modified remove_duplicate_minor_version_books.py, so I'll assume it'll work with the current QA database.  The TEA database has similar problems.

Reply:
I'm focusing on writing the migrations against the production data - since nothing has baked on production yet, I don't think a data-migration is needed for deleting duplicate CompositeModules. It's environment specific, so putting it in a devpos seems the right thing to do to document and control what we're doing. (Hmm, I guess the tea == production actually means we _have_ baked on production …) However, a migration to add the indexes to enable rapid deletion is probably needed. Since this one's getting to be a kitchen-sink anyway, I'll add them, since I have them in my psql history :-)